File: packages/diseno_flutter/lib/organismos/lista_de_movimientos.dart
import 'package:flutter/material.dart';

import '../atomos/monto.dart';
import '../moleculas/fila_de_movimiento.dart';
import '../organismos/movimiento.dart';
import '../tokens/tokens.dart';

/// El extracto: **agrupado por día**, con el neto del día en la cabecera y **un saldo
/// por día** (en un celular el saldo corrido en cada fila es ruido). Perezoso.
class ListaDeMovimientos extends StatelessWidget {
  const ListaDeMovimientos({
    super.key,
    required this.movimientos,
    required this.moneda,
    this.onTap,
  });
  final List<Movimiento> movimientos;
  final String moneda;
  final void Function(Movimiento m)? onTap;

  /// Suma de cadenas del contrato sin pasar por `double`: centavos enteros.
  static String neto(Iterable<String> montos) {
    var centavos = BigInt.zero;
    for (final m in montos) {
      final negativo = m.startsWith('-');
      final partes = (negativo ? m.substring(1) : m).split('.');
      final valor =
          BigInt.parse(partes[0]) * BigInt.from(100) + BigInt.parse(partes[1]);
      centavos += negativo ? -valor : valor;
    }
    final signo = centavos.isNegative ? '-' : '';
    final abs = centavos.abs();
    final enteros = abs ~/ BigInt.from(100);
    final resto = (abs % BigInt.from(100)).toString().padLeft(2, '0');
    return '$signo$enteros.$resto';
  }

  List<(String dia, List<Movimiento>)> get _porDia {
    final grupos = <String, List<Movimiento>>{};
    for (final m in movimientos) {
      grupos.putIfAbsent(m.fechaIso.substring(0, 10), () => []).add(m);
    }
    final dias = grupos.keys.toList()..sort((a, b) => b.compareTo(a));
    return [for (final d in dias) (d, grupos[d]!)];
  }

  @override
  Widget build(BuildContext context) {
    final t = Tokens.of(context);
    final texto = Theme.of(context).textTheme;
    final grupos = _porDia;
    return ListView.builder(
      itemCount: grupos.length,
      itemBuilder: (context, i) {
        final (dia, lista) = grupos[i];
        final saldoDelDia = lista.first.saldoCorrido;
        return Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Container(
              padding: const EdgeInsets.symmetric(
                vertical: Espacio.s2,
                horizontal: Espacio.s1,
              ),
              child: Wrap(
                alignment: WrapAlignment.spaceBetween,
                crossAxisAlignment: WrapCrossAlignment.center,
                spacing: Espacio.s2,
                runSpacing: Espacio.s1,
                children: [
                  Text(
                    _diaLegible(dia),
                    style: texto.labelLarge?.copyWith(color: t.text2),
                  ),
                  Wrap(
                    crossAxisAlignment: WrapCrossAlignment.center,
                    children: [
                      Text(
                        'Neto ',
                        style: texto.bodySmall?.copyWith(color: t.text3),
                      ),
                      Monto(
                        monto: neto(lista.map((m) => m.monto)),
                        moneda: moneda,
                        etiqueta: 'Neto del día',
                        estilo: texto.bodySmall?.copyWith(color: t.text2),
                      ),
                      if (saldoDelDia != null) ...[
                        Text(
                          '  · Saldo ',
                          style: texto.bodySmall?.copyWith(color: t.text3),
                        ),
                        Monto(
                          monto: saldoDelDia,
                          moneda: moneda,
                          etiqueta: 'Saldo al cierre del día',
                          estilo: texto.bodySmall?.copyWith(
                            color: t.brandTexto,
                          ),
                        ),
                      ],
                    ],
                  ),
                ],
              ),
            ),
            for (final m in lista)
              FilaDeMovimiento(
                tipo: m.tipo,
                concepto: m.concepto,
                monto: m.monto,
                moneda: moneda,
                pendiente: m.pendiente,
                onTap: onTap == null ? null : () => onTap!(m),
              ),
            Divider(color: t.border, height: Espacio.s3),
          ],
        );
      },
    );
  }

  static const _meses = [
    'ene',
    'feb',
    'mar',
    'abr',
    'may',
    'jun',
    'jul',
    'ago',
    'sep',
    'oct',
    'nov',
    'dic',
  ];
  static String _diaLegible(String d) {
    final f = DateTime.parse(d);
    return '${f.day} ${_meses[f.month - 1]} ${f.year}';
  }
}

File: packages/ui/src/lista-de-movimientos/lista-de-movimientos.ts
import { ChangeDetectionStrategy, Component, computed, input } from '@angular/core'
import { FilaDeMovimiento, Movimiento } from '../fila-de-movimiento/fila-de-movimiento'
import { Monto } from '../monto/monto'

/**
 * El extracto **agrupado por día** (regla 2), con el neto del día y el saldo al cierre.
 * El neto se suma en centavos con BigInt: nunca con `number`.
 */
@Component({
  selector: 'ap-lista-de-movimientos',
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [FilaDeMovimiento, Monto],
  template: `
    @for (g of grupos(); track g.dia) {
      <section [attr.aria-label]="g.legible">
        <div class="cabecera">
          <h3>{{ g.legible }}</h3>
          <p>Neto <ap-monto [monto]="g.neto" [moneda]="moneda()" etiqueta="Neto del día" /> @if (g.saldo) { · Saldo <ap-monto [monto]="g.saldo" [moneda]="moneda()" etiqueta="Saldo al cierre del día" class="saldo" /> }</p>
        </div>
        @for (m of g.movimientos; track m.id) { <ap-fila-de-movimiento [m]="m" /> }
      </section>
    }
  `,
  styles: `
    .cabecera { display: flex; flex-wrap: wrap; align-items: center; justify-content: space-between; gap: var(--s2); padding: var(--s3) 0 var(--s1); }
    h3 { font-size: .95em; color: var(--text-2); }
    p { margin: 0; color: var(--text-3); font-size: .85em; }
    ap-monto { color: var(--text-2); }
    .saldo { color: var(--brand-texto); }
  `,
})
export class ListaDeMovimientos {
  readonly movimientos = input.required<Movimiento[]>()
  readonly moneda = input.required<string>()
  readonly hoyIso = input.required<string>()
  readonly grupos = computed(() => {
    const porDia = new Map<string, Movimiento[]>()
    for (const m of this.movimientos()) {
      const dia = m.fechaIso.slice(0, 10)
      porDia.set(dia, [...(porDia.get(dia) ?? []), m])
    }
    return [...porDia.entries()]
      .sort(([a], [b]) => (a < b ? 1 : -1))
      .map(([dia, movimientos]) => ({ dia, legible: diaLegible(dia, this.hoyIso().slice(0, 10)), movimientos, neto: neto(movimientos.map((m) => m.monto)), saldo: movimientos[0]?.saldoCorrido }))
  })
}

/** Suma en centavos con BigInt y devuelve la cadena del contrato. */
export function neto(montos: string[]): string {
  let centavos = 0n
  for (const m of montos) {
    const negativo = m.startsWith('-')
    const [entero = '0', dec = '00'] = m.replace('-', '').split('.')
    const c = BigInt(entero) * 100n + BigInt(dec.padEnd(2, '0').slice(0, 2))
    centavos += negativo ? -c : c
  }
  const signo = centavos < 0n ? '-' : ''
  const abs = centavos < 0n ? -centavos : centavos
  return `${signo}${abs / 100n}.${String(abs % 100n).padStart(2, '0')}`
}

export function diaLegible(dia: string, hoy: string): string {
  if (dia === hoy) return 'Hoy'
  const ayer = new Date(new Date(hoy + 'T12:00:00Z').getTime() - 86_400_000).toISOString().slice(0, 10)
  if (dia === ayer) return 'Ayer'
  const [a, m, d] = dia.split('-')
  return `${Number(d)} ${['ene', 'feb', 'mar', 'abr', 'may', 'jun', 'jul', 'ago', 'sep', 'oct', 'nov', 'dic'][Number(m) - 1]} ${a}`
}
